sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp

```diff
@@ -0,0 +1,777 @@
+      bufferSlicePitch, hostSlicePitch, numSyncPointsInWaitList,
+      pSyncPointWaitList, pSyncPoint);
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+    ur_exp_command_buffer_handle_t hCommandBuffer, ur_queue_handle_t hQueue,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
+  UR_ASSERT(hCommandBuffer, UR_RESULT_ERROR_INVALID_COMMAND_BUFFER_EXP);
+
+  std::scoped_lock<ur_shared_mutex> lock(hQueue->Mutex);
+  // Use compute engine rather than copy engine
+  const auto UseCopyEngine = false;
+  auto &QGroup = hQueue->getQueueGroup(UseCopyEngine);
+  uint32_t QueueGroupOrdinal;
+  auto &ZeCommandQueue = QGroup.getZeQueue(&QueueGroupOrdinal);
+
+  ze_fence_handle_t ZeFence;
+  ZeStruct<ze_fence_desc_t> ZeFenceDesc;
+  ur_command_list_ptr_t CommandListPtr;
+
+  ZE2UR_CALL(zeFenceCreate, (ZeCommandQueue, &ZeFenceDesc, &ZeFence));
+  // TODO: Refactor so requiring a map iterator is not required here, currently
+  // required for executeCommandList though.
+  ZeStruct<ze_command_queue_desc_t> ZeQueueDesc;
+  ZeQueueDesc.ordinal = QueueGroupOrdinal;
+  CommandListPtr = hCommandBuffer->CommandListMap.insert(
+      std::pair<ze_command_list_handle_t, pi_command_list_info_t>(
+          hCommandBuffer->ZeCommandList,
+          {ZeFence, false, false, ZeCommandQueue, ZeQueueDesc}));
+
+  // Previous execution will have closed the command list, we need to reopen
+  // it otherwise calling `executeCommandList` will return early.
+  CommandListPtr->second.IsClosed = false;
+  CommandListPtr->second.ZeFenceInUse = true;
+
+  // Create command-list to execute before `CommandListPtr` and will signal
+  // when `EventWaitList` dependencies are complete.
+  ur_command_list_ptr_t WaitCommandList{};
+  if (numEventsInWaitList) {
+    _ur_ze_event_list_t TmpWaitList;
+    if (auto Res = TmpWaitList.createAndRetainUrZeEventList(
+            numEventsInWaitList, phEventWaitList, hQueue, UseCopyEngine))
+      return Res;
+
+    if (auto Res = hQueue->Context->getAvailableCommandList(
+            hQueue, WaitCommandList, false, false))
+      return Res;
+
+    // Update the WaitList of the Wait Event
+    // Events are appended to the WaitList if the WaitList is not empty
+    if (hCommandBuffer->WaitEvent->WaitList.isEmpty())
+      hCommandBuffer->WaitEvent->WaitList = TmpWaitList;
+    else
+      hCommandBuffer->WaitEvent->WaitList.insert(TmpWaitList);
+
+    ZE2UR_CALL(zeCommandListAppendBarrier,
+               (WaitCommandList->first, hCommandBuffer->WaitEvent->ZeEvent,
+                hCommandBuffer->WaitEvent->WaitList.Length,
+                hCommandBuffer->WaitEvent->WaitList.ZeEventList));
+  } else {
+    if (auto Res = hQueue->Context->getAvailableCommandList(
+            hQueue, WaitCommandList, false, false))
+      return Res;
+
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
+               (WaitCommandList->first, hCommandBuffer->WaitEvent->ZeEvent));
```

Comment:
dont understand this. What is WaitEvent and why do we need to append a signal to it when submitting the command buffer?

Reply:
`WaitEvent` is an event that belongs to the command-buffer that is intended to block all the user added commands from running until it is signalled by new prefix command-buffer. This is to solve the problem of representing the wait-list of `urcommandbufferenqueueexp` to a closed command-list.

Reading through this code again, I don't think we ever reset `WaitEvent`  between `urcommandbufferenqueueexp` calls, which maybe we should do in the third suffix command-list. Please let me know your thoughts on that, and/or the solution in general.